### ISSUE TITLE
File.Copy accepts a directory on OSX

### ISSUE THREAD

**Issue by sbomer** [dotnet/corefx#36713]
The documentation for `File.Copy` says:
```
destFileName
    String 

The name of the destination file. This cannot be a directory or an existing file.
```
The windows behavior matches the documentation - we get an exception thrown if the second argument is a directory:
```
Unhandled Exception: System.IO.IOException: The target file 'D:\tmp\cp\b' is a directory, not a file.
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName)
   at cp.Program.Main(String[] args) in D:\tmp\cp\Program.cs:line 10
```

On OSX, passing a directory results in the file being copied into the directory without any errors.

**Comment by danmosemsft** [comment]
@jeremykuhne perhaps we should make the API allow a directory on other platforms?

**Comment by JeremyKuhne** [comment]
Similar to #37121. We should match the legacy behavior (i.e. Windows).

**Comment by davidkaya** [comment]
I will take this one if possible :)

I also noticed that the unit test for this logic false positively passes on macOS because:

```c#
[Fact] 
 public void CopyOntoDirectory() 
 { 
     string testFile = GetTestFilePath(); // Creates path /../TestDirectory/TempFile
     File.Create(testFile).Dispose(); // Creates the file at Path
     Assert.Throws<IOException>(() => Copy(testFile, TestDirectory)); // Tries to copy the file to /../TestDirectory but fails because the file already exists at the path, not because it is a directory
 } 
```

If I remove the `Assert.Throws` and just try to copy it, test throws this `System.IO.IOException : The file '/var/folders/v_/9rh7c5816_gfl6zg4mr8ck8sd9720x/T/File_Copy_str_str_sgluhewm.nv5/CopyOntoDirectory_37_43a79351' already exists.`

So as part of the fix, the unit test also should be changed to something like

```diff
diff --git a/src/System.IO.FileSystem/tests/File/Copy.cs b/src/System.IO.FileSystem/tests/File/Copy.cs
index 1caca2aae1..b677a6e654 100644
--- a/src/System.IO.FileSystem/tests/File/Copy.cs
+++ b/src/System.IO.FileSystem/tests/File/Copy.cs
@@ -35,8 +35,9 @@ namespace System.IO.Tests
         public void CopyOntoDirectory()
         {
             string testFile = GetTestFilePath();
+            string targetTestDirectory = Directory.CreateDirectory(GetTestFilePath()).FullName;
             File.Create(testFile).Dispose();
-            Assert.Throws<IOException>(() => Copy(testFile, TestDirectory));
+            Assert.Throws<IOException>(() => Copy(testFile, targetTestDirectory));
         }
```

**Comment by JeremyKuhne** [comment]
> I will take this one if possible :)

Assigned to you, thanks!

**Comment by Gnbrkm41** [comment]
Seems like it's fixed now via the referencing PR (#37428), shouldn't we close this? @JeremyKuhne

**Comment by JeremyKuhne** [comment]
> shouldn't we close this?

Yes, thanks for catching this!
